File: leetcode/inorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var inorderHelper = function(root, path) {
    if (root === null) return path;
    inorderHelper(root.left, path);
    path.push(root.val);
    inorderHelper(root.right, path);
    return path;
};

var inorderTraversal = function(root) {
    const path = [];
    return inorderHelper(root, path);
};

File: leetcode/inorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {

  // Iterative

  public List<Integer> inorderTraversal(TreeNode root) {
    List<Integer> list = new ArrayList<Integer>();
    Stack<TreeNode> stack = new Stack<TreeNode>();
    TreeNode cur = root;
    while (cur != null || !stack.empty()) {
      while (cur != null) {
        stack.add(cur);
        cur = cur.left;
      }
      cur = stack.pop();
      list.add(cur.val);
      cur = cur.right;
    }
    return list;
  }

  // Recursive

  public List<Integer> inorderTraversal(TreeNode root) {
    List<Integer> list = new ArrayList<Integer>();
    return inorderHelper(root, list);
  }

  public List<Integer> inorderHelper(TreeNode root, List<Integer> list) {
    if (root == null) return list;
    inorderHelper(root.left, list);
    list.add(root.val);
    inorderHelper(root.right, list);
    return list;
  }
}
